The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8425961859/job/23073212369) [(plain)](https://github.com/rust-lang/rust/commit/dba9c26ee49aa284c231d7e6a54cb0047d244fb1/checks/23073212369/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#17 exporting to docker image format
#17 sending tarball 28.7s done
#17 DONE 33.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

---- [ui] tests/ui/consts/const-eval/const_panic.rs stdout ----
diff of stderr:

+ WARN rustc_errors::emitter Invalid span $SRC_DIR/core/src/panic.rs:LL:COL (#34), error=SourceNotAvailable { filename: Real(Remapped { local_path: None, virtual_name: "$SRC_DIR/core/src/panic.rs" }) }
+ error[E0425]: cannot find function `panic_str` in module `$crate::panicking`
+    |
+    |
+ LL | const W_CORE: () = core::panic!(MSG);
+   --> $SRC_DIR/core/src/panicking.rs:LL:COL
+    |
+    = note: similarly named function `panic_fmt` defined here
+    |
---
80 
81 error[E0080]: evaluation of constant value failed
-   --> $DIR/const_panic.rs:36:20
-    |
- LL | const W_CORE: () = core::panic!(MSG);
-    |                    ^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'hello', $DIR/const_panic.rs:36:20
-    = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `core::panic` (in Nightly builds, run with -Z macro-backtrace for more info)
- 
- error[E0080]: evaluation of constant value failed
90   --> $DIR/const_panic.rs:39:21
90   --> $DIR/const_panic.rs:39:21
91    |
92 LL | const W2_CORE: () = core::panic!("{}", MSG);
96 
97 error: aborting due to 12 previous errors
98 
- For more information about this error, try `rustc --explain E0080`.
---
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/consts/const-eval/const_panic.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-eval/const_panic" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-eval/const_panic/auxiliary"
stdout: none
--- stderr -------------------------------
WARN rustc_errors::emitter Invalid span /rustc/FAKE_PREFIX/library/core/src/panic.rs:32:28: 32:37 (#34), error=SourceNotAvailable { filename: Real(Remapped { local_path: None, virtual_name: "/rustc/FAKE_PREFIX/library/core/src/panic.rs" }) }
error[E0425]: cannot find function `panic_str` in module `$crate::panicking`
   |
   |
LL | const W_CORE: () = core::panic!(MSG);
  --> /rustc/FAKE_PREFIX/library/core/src/panicking.rs:52:1
   |
   = note: similarly named function `panic_fmt` defined here
   |
   |
   = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `core::panic` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0080]: evaluation of constant value failed
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:6:15
   |
LL | const Z: () = std::panic!("cheese");
   |               ^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'cheese', /checkout/tests/ui/consts/const-eval/const_panic.rs:6:15
   = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `std::panic` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0080]: evaluation of constant value failed
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:9:16
---

error[E0080]: evaluation of constant value failed
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:18:15
   |
LL | const W: () = std::panic!(MSG);
   |               ^^^^^^^^^^^^^^^^ the evaluated program panicked at 'hello', /checkout/tests/ui/consts/const-eval/const_panic.rs:18:15
   = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `std::panic` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0080]: evaluation of constant value failed
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:21:16
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:21:16
   |
LL | const W2: () = std::panic!("{}", MSG);
   |                ^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'hello', /checkout/tests/ui/consts/const-eval/const_panic.rs:21:16
   = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `std::panic` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0080]: evaluation of constant value failed
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:24:20
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:24:20
   |
LL | const Z_CORE: () = core::panic!("cheese");
   |                    ^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'cheese', /checkout/tests/ui/consts/const-eval/const_panic.rs:24:20
   = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `core::panic` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0080]: evaluation of constant value failed
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:27:21
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:27:21
   |
LL | const Z2_CORE: () = core::panic!();
   |
   = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `core::panic` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0080]: evaluation of constant value failed
error[E0080]: evaluation of constant value failed
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:30:20
   |
LL | const Y_CORE: () = core::unreachable!();
   |                    ^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'internal error: entered unreachable code', /checkout/tests/ui/consts/const-eval/const_panic.rs:30:20
   = note: this error originates in the macro `$crate::panic::unreachable_2015` which comes from the expansion of the macro `core::unreachable` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0080]: evaluation of constant value failed
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:33:20
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:33:20
   |
LL | const X_CORE: () = core::unimplemented!();
   |                    ^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'not implemented', /checkout/tests/ui/consts/const-eval/const_panic.rs:33:20
   = note: this error originates in the macro `core::unimplemented` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0080]: evaluation of constant value failed
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:39:21
##[error]  --> /checkout/tests/ui/consts/const-eval/const_panic.rs:39:21
   |
LL | const W2_CORE: () = core::panic!("{}", MSG);
   |                     ^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'hello', /checkout/tests/ui/consts/const-eval/const_panic.rs:39:21
   = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `core::panic` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 12 previous errors

---
Build completed unsuccessfully in 0:14:09
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/non-fmt-panic.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/non-fmt-panic" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/non-fmt-panic/auxiliary"
stdout: none
--- stderr -------------------------------
WARN rustc_errors::emitter Invalid span /rustc/FAKE_PREFIX/library/core/src/panic.rs:32:28: 32:37 (#57), error=SourceNotAvailable { filename: Real(Remapped { local_path: None, virtual_name: "/rustc/FAKE_PREFIX/library/core/src/panic.rs" }) }
error[E0425]: cannot find function `panic_str` in module `$crate::panicking`
   |
   |
LL |     core::panic!(&*"abc"); //~ WARN panic message is not a string literal
  --> /rustc/FAKE_PREFIX/library/core/src/panicking.rs:52:1
   |
   = note: similarly named function `panic_fmt` defined here
   |
```

</details>

--- a/compiler/rustc_lint/src/non_fmt_panic.rs
+++ b/compiler/rustc_lint/src/non_fmt_panic.rs
@@ -53,7 +53,7 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
 
                 if Some(def_id) == cx.tcx.lang_items().begin_panic_fn()
                     || Some(def_id) == cx.tcx.lang_items().panic_fn()
-                    || f_diagnostic_name == Some(sym::panic_str)
+                    || f_diagnostic_name == Some(sym::panic_str_2015)
                 {
                     if let Some(id) = f.span.ctxt().outer_expn_data().macro_def_id {
                         if matches!(

--- a/compiler/rustc_span/src/symbol.rs
+++ b/compiler/rustc_span/src/symbol.rs
@@ -1306,7 +1306,7 @@ symbols! {
         panic_misaligned_pointer_dereference,
         panic_nounwind,
         panic_runtime,
-        panic_str,
+        panic_str_2015,
         panic_unwind,
         panicking,
         param_attrs,

--- a/library/core/src/option.rs
+++ b/library/core/src/option.rs
@@ -554,7 +554,7 @@
 #![stable(feature = "rust1", since = "1.0.0")]
 
 use crate::iter::{self, FusedIterator, TrustedLen};
-use crate::panicking::{panic, panic_str};
+use crate::panicking::{panic, panic_display};
 use crate::pin::Pin;
 use crate::{
     cmp, convert, hint, mem,
@@ -1991,7 +1991,7 @@ const fn unwrap_failed() -> ! {
 #[track_caller]
 #[rustc_const_unstable(feature = "const_option", issue = "67441")]
 const fn expect_failed(msg: &str) -> ! {
-    panic_str(msg)
+    panic_display(&msg)
 }
 
 /////////////////////////////////////////////////////////////////////////////

--- a/library/core/src/panic.rs
+++ b/library/core/src/panic.rs
@@ -27,9 +27,9 @@ pub macro panic_2015 {
     ($msg:literal $(,)?) => (
         $crate::panicking::panic($msg)
     ),
-    // Use `panic_str` instead of `panic_display::<&str>` for non_fmt_panic lint.
+    // Use `panic_str_2015` instead of `panic_display::<&str>` for non_fmt_panic lint.
     ($msg:expr $(,)?) => ({
-        $crate::panicking::panic_str($msg);
+        $crate::panicking::panic_str_2015($msg);
     }),
     // Special-case the single-argument case for const_panic.
     ("{}", $arg:expr $(,)?) => ({

--- a/library/core/src/panicking.rs
+++ b/library/core/src/panicking.rs
@@ -124,8 +124,8 @@ pub const fn panic_nounwind_fmt(fmt: fmt::Arguments<'_>, force_no_backtrace: boo
 // above.
 
 /// The underlying implementation of core's `panic!` macro when no formatting is used.
-// never inline unless panic_immediate_abort to avoid code
-// bloat at the call sites as much as possible
+// Never inline unless panic_immediate_abort to avoid code
+// bloat at the call sites as much as possible.
 #[cfg_attr(not(feature = "panic_immediate_abort"), inline(never), cold)]
 #[cfg_attr(feature = "panic_immediate_abort", inline)]
 #[track_caller]
@@ -138,6 +138,11 @@ pub const fn panic(expr: &'static str) -> ! {
     // truncation and padding (even though none is used here). Using
     // Arguments::new_const may allow the compiler to omit Formatter::pad from the
     // output binary, saving up to a few kilobytes.
+    // However, this optimization only works for `'static` strings: `new_const` also makes this
+    // message return `Some` from `Arguments::as_str`, which means it can become part of the panic
+    // payload without any allocation or copying. Shorter-lived strings would become invalid as
+    // stack frames get popped during unwinding, and couldn't be directly referenced from the
+    // payload.
     panic_fmt(fmt::Arguments::new_const(&[expr]));
 }
 
@@ -160,14 +165,6 @@ pub fn panic_nounwind_nobacktrace(expr: &'static str) -> ! {
     panic_nounwind_fmt(fmt::Arguments::new_const(&[expr]), /* force_no_backtrace */ true);
 }
 
-#[inline]
-#[track_caller]
-#[rustc_diagnostic_item = "panic_str"]
-#[rustc_const_unstable(feature = "panic_internals", issue = "none")]
-pub const fn panic_str(expr: &str) -> ! {
-    panic_display(&expr);
-}
-
 #[track_caller]
 #[cfg_attr(not(feature = "panic_immediate_abort"), inline(never), cold)]
 #[cfg_attr(feature = "panic_immediate_abort", inline)]
@@ -181,6 +178,16 @@ pub const fn panic_explicit() -> ! {
 #[rustc_diagnostic_item = "unreachable_display"] // needed for `non-fmt-panics` lint
 pub fn unreachable_display<T: fmt::Display>(x: &T) -> ! {
     panic_fmt(format_args!("internal error: entered unreachable code: {}", *x));
+}
+
+/// This exists solely for the 2015 edition `panic!` macro to trigger
+/// a lint on `panic!(my_str_variable);`.
+#[inline]
+#[track_caller]
+#[rustc_diagnostic_item = "panic_str_2015"]
+#[rustc_const_unstable(feature = "panic_internals", issue = "none")]
+pub const fn panic_str_2015(expr: &str) -> ! {
+    panic_display(&expr);
 }
 
 #[inline]

